dex/src/main/scala/com/wavesplatform/dex/db/LocalQueueStore.scala

```diff
@@ -1,38 +1,56 @@
 package com.wavesplatform.dex.db
 
+import cats.Monad
+import cats.implicits._
+
 import java.util.concurrent.ConcurrentLinkedQueue
 import java.util.concurrent.atomic.AtomicLong
-
 import com.google.common.primitives.{Longs, Shorts}
 import com.wavesplatform.dex.db.DbKeys._
-import com.wavesplatform.dex.db.leveldb.{DBExt, ReadOnlyDB}
+import com.wavesplatform.dex.db.leveldb.Key
 import com.wavesplatform.dex.queue.{ValidatedCommand, ValidatedCommandWithMeta}
-import org.iq80.leveldb.{DB, ReadOptions}
+import com.wavesplatform.dex.db.leveldb.LevelDb
+
+trait AbstractLocalQueueStore[F[_]] {
+
+  def enqueue(command: ValidatedCommand, timestamp: Long): F[ValidatedCommandWithMeta.Offset]
+
+  def getFrom(offset: ValidatedCommandWithMeta.Offset, maxElements: Int): F[Vector[ValidatedCommandWithMeta]]
+
+  def oldestOffset: F[Option[ValidatedCommandWithMeta.Offset]]
+
+  def newestOffset: F[Option[ValidatedCommandWithMeta.Offset]]
+
+  def dropUntil(offset: ValidatedCommandWithMeta.Offset): F[Unit]
+
+}
 
-class LocalQueueStore(db: DB) {
+class LocalQueueStore[F[_]](levelDb: LevelDb[F])(implicit fa: Monad[F]) extends AbstractLocalQueueStore[F]{
 
-  private val newestIdx = new AtomicLong(db.get(lqNewestIdx))
+  private val newestIdx = levelDb.get(lqNewestIdx).map(s => new AtomicLong(s))
   private val inMemQueue = new ConcurrentLinkedQueue[ValidatedCommandWithMeta]
   private var startInMemOffset = Option.empty[ValidatedCommandWithMeta.Offset]
 
-  def enqueue(command: ValidatedCommand, timestamp: Long): ValidatedCommandWithMeta.Offset = {
-    val offset = newestIdx.incrementAndGet()
-    val eventKey = lpqElement(offset)
+  override def enqueue(command: ValidatedCommand, timestamp: Long): F[ValidatedCommandWithMeta.Offset] = {
+    newestIdx.flatMap { oldOffset =>
+      val offset = oldOffset.incrementAndGet()
+      val eventKey: Key[Option[ValidatedCommandWithMeta]] = lpqElement(offset)
 
-    val x = ValidatedCommandWithMeta(offset, timestamp, command)
-    db.readWrite { rw =>
-      rw.put(eventKey, Some(x))
-      rw.put(lqNewestIdx, offset)
+      val x = ValidatedCommandWithMeta(offset, timestamp, command)
+      levelDb.readWrite { rw =>
+        rw.put(eventKey, Some(x))
+        rw.put(lqNewestIdx, offset)
+      }.map { _ =>
+        inMemQueue.add(x)
```

Comment:
Probably there are concurrent issues when `enqueue` is working concurrently with `getFrom`. Could you add a TODO here?
```scala
// TODO Probably concurrent work issues
```